---
title: "Advanced Spatial Analysis"
author:
- name: Paula Nieto
  affiliation: 
  - Centro Nacional de Análisis Genómico (CNAG)
  email: paula.nieto@cnag.crg.eu
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

# Analysis of Spatial Data

We are going to analyze a breast cancer section from [10x Genomicsis](https://www.10xgenomics.com/) Visium Gene Expression platform.

For the analysis we will mainly use the R package [Seurat](https://satijalab.org/seurat/index.html).

## Download Data

You can download the data from [this website](https://www.10xgenomics.com/resources/datasets/human-breast-cancer-block-a-section-1-1-standard-1-1-0) or using `curl`, as shown below. We do not need to do it because it is already in the `data` folder.

```{bash, eval = FALSE}
# download files
curl -O https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Breast_Cancer_Block_A_Section_1/V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5
curl -O https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Breast_Cancer_Block_A_Section_1/V1_Breast_Cancer_Block_A_Section_1_spatial.tar.gz
```

## Libraries

These are the libraries we are going to use:

```{r libraries, echo = FALSE}
library(tidyverse)
library(Seurat)
library(patchwork)
```


```{r seed}
# set seed for reproducibility purposes
set.seed(12345)

# create palette for the cell types
cell_type_palette <- as.vector(pals::polychrome())
```

## Load data with Seurat

```{r}
se_obj <- Load10X_Spatial(
  data.dir = "data"
)
```

The visium data from 10x consists is stored in a Seurat object as:

* A spot x gene expression matrix (similar to the cell x gene matrix)

```{r}
se_obj[["Spatial"]][1:5, 1:5]
```

* Image of the tissue slice (obtained from staining during sample processing in the lab)

```{r}
SpatialPlot(
  se_obj, 
  pt.size = 0
  ) + 
  NoLegend()
```


* Scaling factors that relate the original high resolution image to the lower resolution image used here for visualization.

```{r}
se_obj@images$slice1@scale.factors
```

## Quality control

As with single-cell objects, we have some important features that we can use to filter out bad quality spots.

* nCount_Spatial: number of UMIs per spot

```{r fig.width=12}
plot_1 <- VlnPlot(
  se_obj, 
  features = "nCount_Spatial", 
  pt.size = 0.1
  ) + 
  NoLegend()

plot_2 <- SpatialFeaturePlot(
  se_obj, 
  features = "nCount_Spatial"
  )

plot_1 | plot_2
```

The variability in the distribution of UMIs is related to the tissue architecture, i.e. tumoral regions are more densely packked and thus contain spots with higher counts.

* nFeature_Spatial: number of genes per spot

```{r fig.width=10}
plot_1 <- VlnPlot(
  se_obj, 
  features = "nFeature_Spatial", 
  pt.size = 0.1
  ) + 
  NoLegend()

plot_2 <- SpatialFeaturePlot(
  se_obj, 
  features = "nFeature_Spatial"
  )

plot_1 | plot_2
```

Here we can also see how the number of genes per spot correlates with the structure of the tissue

* mt.content and rb.content: mitochondrial and ribosomal content per spot, respectively

We have to compute this two values by calculating the percentage of reads per spot that belong to mitochondrial/ribosomal genes.

```{r}
# Mitochondrial content
se_obj[["mt.content"]] <- PercentageFeatureSet(
  object = se_obj,
  pattern = "^MT-")

summary(se_obj[["mt.content"]])
```

```{r fig.width=10}
# Ral contentibosomal
se_obj[["rb.content"]] <- PercentageFeatureSet(
  object = se_obj,
  pattern = "^RPL|^RPS")

summary(se_obj[["rb.content"]])
```

```{r fig.width=10, fig.height=9}
SpatialFeaturePlot(
  se_obj, 
  features = c("mt.content", "rb.content")
  )
```

In the case of spatial data, high mitochondrial content is not an indicator of bad quality spots and thus should not be a criterion to filter out spots.

We are going to filter out genes that have no expression in the tissue.

```{r}
table(rowSums(as.matrix(se_obj@assays$Spatial@counts)) == 0)
keep_genes <- rowSums(as.matrix(se_obj@assays$Spatial@counts)) != 0
se_obj <- se_obj[keep_genes, ]
```

Furthermore, we will filter out spots with very low number of counts (< 500) before proceeding with the downstream analysis.

```{r}
se_obj <- subset(
  se_obj,
  subset = nCount_Spatial > 500
)
```

## Preprocessing

Similar to single-cell datasets, preprocessing for spatial data requires normalizing, finding variable features and scaling the counts.

```{r}
se_obj <- NormalizeData(se_obj)
se_obj <- FindVariableFeatures(se_obj)
se_obj <- ScaleData(se_obj)
```

Prior to clustering, we perform dimensionality reduction.

```{r}
se_obj <- RunPCA(se_obj)
se_obj <- RunUMAP(
  se_obj,
  reduction = "pca",
  dims = 1:20
)
```

## Clustering and visualization

```{r}
se_obj <- FindNeighbors(
  se_obj, 
  reduction = "pca", 
  dims = 1:20
  )

se_obj <- FindClusters(
  se_obj,
  resolution = 0.3
  )
```

```{r fig.width=10}
plot_1 <- DimPlot(
  se_obj, 
  label = TRUE
  )

plot_2 <- SpatialDimPlot(
  se_obj, 
) +
  NoLegend()

plot_1 + plot_2
```

## Gene expression and annotation

ESR1 and ERBB2 are the most common mutations in breast cancer, so one way of annotating the tissue is by looking at ESR1 and ERBB2 positive/negative regions.

```{r}
SpatialFeaturePlot(
  se_obj, 
  features = c("ESR1", "ERBB2"), 
  alpha = c(0.1, 1)
  )
```

## Deconvolution

We are going to use [SPOTlight](https://academic.oup.com/nar/article/49/9/e50/6129341?login=true) in conjunction with a subset of the [Tumor Immune Cell Atlas](https://genome.cshlp.org/content/31/10/1913) to deconvolute our spots and map immune populations to our tumor section.

```{r}
# load library
library(SPOTlight)
```

Load downsampled version of the Tumor Immune Cell Atlas and explore the metadata we have. We are going to use annotation level 1 for the deconvolution (`lv1_annot`).

```{r}
atlas <- readRDS("data/TICAtlas_downsample.rds")
head(atlas@meta.data)
```

First of all we need to compute the markers for the cell types using Seurat's funtion `FindAllMarkers`. 

```{r eval = FALSE}
Idents(atlas) <- "lv1_annot"
  
sc_markers <- FindAllMarkers(
  object = atlas,
  assay = "RNA",
  slot = "data",
  only.pos = TRUE
  )

sc_markers <- filter(
  sc_markers,
  avg_log2FC >= 0.7
  )

sc_markers <- filter(
  sc_markers,
  str_detect(string = gene,
             pattern = "^Rps|^Rpl|^Mt-",
             negate = TRUE
             )
  )
saveRDS(sc_markers, "R_obj/filtered_atlas_markers.rds")
```

```{r}
sc_markers <- readRDS("R_obj/filtered_atlas_markers.rds")
sc_markers %>% 
  group_by(cluster) %>% 
  top_n(5, wt = avg_log2FC)
```

Run the deconvolution using the scRNAseq atlas and the spatial transcriptomics data.

```{r eval = FALSE}
decon_mtrx_ls <- spotlight_deconvolution(
  se_sc = atlas,
  counts_spatial = se_obj@assays$Spatial@counts,
  clust_vr = "lv1_annot",
  cluster_markers = sc_markers,
  hvg = 0,
  min_cont = 0,
  assay = "RNA",
  slot = "counts")

saveRDS(decon_mtrx_ls, "R_obj/deconvolution_matrix.rds")
```

```{r}
decon_mtrx_ls <- readRDS("R_obj/deconvolution_matrix.rds")
```

Before even looking at the decomposed spots we can gain insight on how well the model performed by looking at the topic profiles for the cell types.

```{r}
nmf_mod <- decon_mtrx_ls[[1]]
decon_mtrx <- decon_mtrx_ls[[2]]

rownames(decon_mtrx) <- colnames(se_obj)

# info on the NMF model
nmf_mod[[1]]

# deconvolution matrix
head(decon_mtrx)
```

Look at how specific the topic profiles are for each cell type.

```{r fig.height=10, fig.width=10}
h <- NMF::coef(nmf_mod[[1]])
rownames(h) <- paste("Topic", 1:nrow(h), sep = "_")
topic_profile_plts <- SPOTlight::dot_plot_profiles_fun(
  h = h,
  train_cell_clust = nmf_mod[[2]])

topic_profile_plts[[2]]
```

Let's now visualize how the deconvoluted spots on the the Seurat object.

```{r}
# keep on
decon_mtrx <- decon_mtrx_ls[[2]]
decon_mtrx <- decon_mtrx[, colnames(decon_mtrx) != "res_ss"]
decon_mtrx <- decon_mtrx[, !is.na(colnames(decon_mtrx))]
```


```{r eval = FALSE}
se_obj@meta.data <- cbind(se_obj@meta.data, decon_mtrx)
saveRDS(se_obj, "R_obj/breast_slide_deconvoluted.rds")
```

We have an object with the deconvolution information already added:

```{r}
se_obj <- readRDS("R_obj/breast_slide_deconvoluted.rds")
```


```{r}
ct <- colnames(decon_mtrx)

scatterpie_plot(
    se_obj = se_obj,
    cell_types_all = ct,
    pie_scale = 0.3
    ) +
  coord_fixed(ratio = 1) +
  scale_fill_manual(values = cell_type_palette)
```

Look at the spatial scatterpie by looking at cell types which are not present throughout the entire tissue

```{r}
# keep only cell types that are present in less than 80% of the spots
keep_0.8 <- colSums(se_obj@meta.data[, ct] > 0) < 0.8 * ncol(se_obj)
# but not those that were not found on the tissue
keep_g0 <- colSums(se_obj@meta.data[, ct] > 0) > 0

# select cell types fullfiling the conditions
ct_var <- colnames(se_obj@meta.data[, ct])[keep_0.8 & keep_g0]

ct_var
```


```{r}
scatterpie_plot(
  se_obj = se_obj,
  cell_types_all = ct_var,
  pie_scale = 0.3
  ) +
  coord_fixed(ratio = 1) +
  scale_fill_manual(values = cell_type_palette)
```

Lets now see the frequencies of selected cell types across the regions in the tissue. We will focus on cytotoxic T cells and macrophages as they can give an idea of how infiltrated the tumor can be. 

```{r}
VlnPlot(
  se_obj,
  features = c("CD8.terminally.exhausted", "TAMs.proinflamatory"),
  group.by = "seurat_clusters"
)
```

```{r}
plot_1 <- SpatialPlot(
  se_obj, 
  pt.size = 0
  ) + 
  NoLegend()

plot_2 <- SpatialDimPlot(
  se_obj, 
  group.by = "seurat_clusters"
)

plot_1 + plot_2
```

We can see that in region of cluster 2 we have a presence of both exhausted T cells and proinflamatory macrophages.

*** 

<a href="#top">Back to top</a>

```{r}
sessionInfo()
```
